server/src/main/java/org/elasticsearch/common/util/concurrent/AbstractThrottledTaskRunner.java

```diff
@@ -155,4 +156,30 @@ int runningTasks() {
         return runningTasks.get();
     }
 
+    /**
+     * Eagerly pull tasks from the queue and execute them on this thread. This must only be used if the tasks in the queue are all
+     * synchronous, i.e. they release their ref before returning from {@code onResponse()}.
+     */
+    public void runSyncTasksEagerly() {
+        final AtomicBoolean isDone = new AtomicBoolean(true);
+        final Releasable ref = () -> isDone.set(true);
+        ActionListener<Releasable> task;
+        while ((task = tasks.poll()) != null) {
+            isDone.set(false);
+            try {
+                logger.trace("[{}] eagerly running task {}", taskRunnerName, task);
+                task.onResponse(ref);
+            } catch (Exception e) {
+                logger.error(Strings.format("[%s] task %s failed", taskRunnerName, task), e);
+                assert false : e;
+                task.onFailure(e);
+                return;
+            }
+            if (isDone.get() == false) {
+                logger.error("runSyncTasksEagerly() was called on a queue [{}] containing an async task: [{}]", taskRunnerName, task);
+                assert false;
```

Comment:
Yes indeed but that doesn't matter to us really. The `ref` is just so that the `AbstractThrottledTaskRunner` can track the relevant activities to completion. If we did fork an untracked task then presumably we're tracking it somewhere else.